processor/src/tests/u32_ops.rs

```diff
@@ -163,63 +163,650 @@ fn u32split() {
+    let asm_op = "u32mul.full";
+
+    // should fail if a >= 2^32
+    test_execution_failure(asm_op, &[0, U32_BOUND], "FailedAssertion");
+
+    // should fail if b >= 2^32
+    test_execution_failure(asm_op, &[U32_BOUND, 0], "FailedAssertion");
+}
+
+#[test]
+fn u32mul_unsafe() {
+    let asm_op = "u32mul.unsafe";
+
+    // should push c = (a * b) % 2^32 onto the stack
+    // should push d = (a * b) / 2^32 onto the stack
+    test_mul_full(asm_op);
+
+    // should not fail when inputs are out of bounds
+    test_unsafe_execution(asm_op, 2);
 }
 
 #[test]
-#[ignore = "unimplemented"]
 fn u32madd() {
-    unimplemented!();
+    let asm_op = "u32madd";
+
+    // should push d = (a * b + c) % 2^32 onto the stack
+    // should push e = (a * b + c) / 2^32 onto the stack
+    test_madd(asm_op);
+}
+
+#[test]
+fn u32madd_fail() {
+    let asm_op = "u32madd";
+
+    // should fail if a >= 2^32
+    test_execution_failure(asm_op, &[0, U32_BOUND, 0], "FailedAssertion");
+
+    // should fail if b >= 2^32
+    test_execution_failure(asm_op, &[U32_BOUND, 0, 0], "FailedAssertion");
+
+    // should fail if c  >= 2^32
+    test_execution_failure(asm_op, &[0, 0, U32_BOUND], "FailedAssertion");
+}
+
+#[test]
+fn u32madd_unsafe() {
+    let asm_op = "u32madd.unsafe";
+
+    // should push d = (a * b + c) % 2^32 onto the stack
+    // should push e = (a * b + c) / 2^32 onto the stack
+    test_madd(asm_op);
+
+    // should not fail when inputs are out of bounds
+    test_unsafe_execution(asm_op, 3);
 }
 
 #[test]
-#[ignore = "unimplemented"]
+// issue: https://github.com/maticnetwork/miden/issues/94
 fn u32div() {
-    unimplemented!();
+    let asm_op = "u32div";
+
+    // --- simple cases ---------------------------------------------------------------------------
+    test_execution(asm_op, &[1, 0], &[0]);
+    // division with no remainder
+    test_execution(asm_op, &[1, 2], &[2]);
+    // division with remainder
+    test_execution(asm_op, &[2, 1], &[0]);
+
+    // --- random u32 values ----------------------------------------------------------------------
+    let a = rand_value::<u64>() as u32;
+    let b = rand_value::<u64>() as u32;
+    let expected = a / b;
+    test_execution(asm_op, &[b as u64, a as u64], &[expected as u64]);
+
+    // --- test that the rest of the stack isn't affected -----------------------------------------
+    let c = rand_value::<u64>();
+    test_execution(asm_op, &[b as u64, a as u64, c], &[expected as u64, c]);
+}
+
+#[test]
+fn u32div_fail() {
+    let asm_op = "u32div";
+
+    // should fail if a >= 2^32
+    test_execution_failure(asm_op, &[1, U32_BOUND], "FailedAssertion");
+
+    // should fail if b >= 2^32
+    test_execution_failure(asm_op, &[U32_BOUND, 1], "FailedAssertion");
+}
+
+#[test]
+#[should_panic = "divide by zero"]
+fn u32div_panic() {
+    let script = compile("begin u32div end");
+    let inputs = build_inputs(&[0, 1]);
+
+    // should panic if b = 0
+    execute(&script, &inputs).unwrap();
+}
```

Comment:
This should probably return an error rather than panicking. I'll create a separate issue for this.

Reply:
I wondered about this. Doing an error instead sounds great